src/main/java/com/google/api/codegen/transformer/go/GoModelTypeNameConverter.java

```diff
@@ -0,0 +1,166 @@
+
+import com.google.api.codegen.transformer.ModelTypeNameConverter;
+import com.google.api.codegen.util.TypedValue;
+import com.google.api.codegen.util.TypeName;
+import com.google.api.tools.framework.model.ProtoElement;
+import com.google.api.tools.framework.model.TypeRef;
+import com.google.common.collect.ImmutableMap;
+import com.google.protobuf.DescriptorProtos.FieldDescriptorProto.Type;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+public class GoModelTypeNameConverter implements ModelTypeNameConverter {
+
+  /**
+   * The import path for generated pb.go files for core-proto files.
+   */
+  private static final String CORE_PROTO_BASE = "google.golang.org/genproto/";
+
+  private static final ImmutableMap<Type, String> PRIMITIVE_TYPE_MAP =
+      ImmutableMap.<Type, String>builder()
+          .put(Type.TYPE_BOOL, "bool")
+          .put(Type.TYPE_DOUBLE, "float64")
+          .put(Type.TYPE_FLOAT, "float32")
+          .put(Type.TYPE_INT64, "int64")
+          .put(Type.TYPE_UINT64, "uint64")
+          .put(Type.TYPE_SINT64, "int64")
+          .put(Type.TYPE_FIXED64, "int64")
+          .put(Type.TYPE_SFIXED64, "int64")
+          .put(Type.TYPE_INT32, "int32")
+          .put(Type.TYPE_UINT32, "uint32")
+          .put(Type.TYPE_SINT32, "int32")
+          .put(Type.TYPE_FIXED32, "int32")
+          .put(Type.TYPE_SFIXED32, "int32")
+          .put(Type.TYPE_STRING, "string")
+          .put(Type.TYPE_BYTES, "[]byte")
+          .build();
+
+  @Override
+  public TypeName getTypeName(TypeRef type) {
+    if (type.isMap()) {
+      TypeName keyTypeName = getTypeNameForElementType(type.getMapKeyField().getType());
+      TypeName valueTypeName = getTypeNameForElementType(type.getMapValueField().getType());
+      return new TypeName("", "", "map[%i]%i", keyTypeName, valueTypeName);
+    } else if (type.isRepeated()) {
+      TypeName elementTypeName = getTypeNameForElementType(type);
+      return new TypeName("", "", "[]%i", elementTypeName);
+    } else {
+      return getTypeNameForElementType(type);
+    }
+  }
+
+  @Override
+  public TypeName getTypeNameForElementType(TypeRef type) {
+    String primitiveTypeName = PRIMITIVE_TYPE_MAP.get(type.getKind());
+    if (primitiveTypeName != null) {
+      return new TypeName(primitiveTypeName);
+    }
+    switch (type.getKind()) {
+      case TYPE_MESSAGE:
+        return getTypeName(type.getMessageType());
+      case TYPE_ENUM:
+        return getTypeName(type.getEnumType());
+      default:
+        throw new IllegalArgumentException("unknown type kind: " + type.getKind());
+    }
+  }
+
+  @Override
+  public TypeName getTypeName(ProtoElement elem) {
+    String pkgName = localPackageName(elem);
+    return new TypeName(pkgName + "." + elem.getSimpleName(), elem.getSimpleName());
+  }
+
+  /**
+   * Returns the local name used in Go files for the package of the proto.
+   */
+  public static String localPackageName(ProtoElement proto) {
```

Comment:
Well, more than simply the `static` keyword; everything else is accessed through the ModelTypeNameConverter interface, whereas this function needs to be accessed directly on GoModelTypeNameConverter, thus bypassing the abstraction.